staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -629,6 +640,18 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		c.watcherIdx++
 	}()
 
+	if !addedWatcher {
+		// For safety we may close the watcher, but it wasn't yet started at this point.
+		watcher.Stop()
+
+		// To match the uncached watch implementation, once we have passed authn/authz/admission,
+		// and successfully parsed a resource version, other errors must fail with a watch event of type ERROR,
+		// rather than a directly returned error.
+		// For consistency with waiting for readingess, we also return service unavailable error.
+		err = errors.NewServiceUnavailable("apiserver cacher is reinitializing")
```

Comment:
is service unavailable correct? what action will the client take in response? do we want the client to get back an error, or just to get the watch closed?

Reply:
This is consistent with what we're returning here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L528
[and conceptually it seems fine - the service is temporarily unavailable]

Regarding handling the response - it obviously depends on the client. But the typical client using reflector will hit this part:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L432-L450

Given this error is not catched, it will cause the client to relist.

Reply:
That being said - this really only happens as a race, so forcing the client to relist in this case doesn't sound terrible.

Reply:
this isn't exactly an error condition, it's a termination of the watch with no events, right?

what does `terminateAllWatchers()` send to established watchers? an error event or just a close? should we make this code path match that behavior?

Reply:
> what does terminateAllWatchers() send to established watchers?

answering my own question, it just closes their done channel

wondering if we should return a `newImmediateCloseWatcher()` here that returns a watch.Interface implementation with a pre-closed result channel and no events

Reply:
> this isn't exactly an error condition, it's a termination of the watch with no events, right?

OK - fair. I agree we don't need to return an error.
The `newImmediateCloseWatcher` makes sense to me - let me change that.